### PR TITLE
Fix item loaded with charges but cannot have any error

### LINKED ISSUE #41821: Object was loaded with charges, but can not have any!
# Describe the bug

While playing the game (version 15a37481, built on Linux64 Mint 19), I received the error:
DEBUG: Item acoustic_guitar was loaded with charges, but can not have any!
FUNCTION: void item::io(Archives&) [with Archive = io::JsonObjectInputArchive]
FILE: src/savegame_json.cpp
LINE: 2390

I believe this occurred in the middle of an auto-save event.

At the time the save began, I may have been walking. There were no acoustic guitars in the room, as far as I can tell, and I was not interacting with one at the time. 

# Steps To Reproduce

The guitar should not be loaded with, or maybe should not bother checking for, charges, I guess. Because: it can not have any!

However, it is not limited to guitar. See context below, and:

After saving and quitting (I tried to generate a game report and although it said it did, it did not produce an object in the config folder, despite multiple attempts), I started the game again, and this debug.log file was created:

> -----------------------------------------
> 19:29:30.269 : Starting log.
> 19:29:30.269 INFO : Cataclysm DDA version 15a37481
> 19:29:30.269 INFO : [main] C locale set to en_US.UTF-8
> 19:29:30.269 INFO : [main] C++ locale set to en_US.UTF-8
> 19:29:30.272 INFO : [options] C locale set to en_US.UTF-8
> 19:29:30.272 INFO : [options] C++ locale set to en_US.UTF-8
> 19:29:30.503 WARNING : opendir [./mods/] failed with "No such file or directory".
> 19:29:37.685 WARNING : opendir [./save/Harmony/mods] failed with "No such file or directory".

Furthermore, I got another LINE 2390 error upon entering the game:
```
Item handflare was loaded with charges, but can not have any!
void item::io(Archive&) [with Archive = io::JsonObjectInputArchive]
src/savegame_json.cpp
LINE 2390
```

# Screenshots

# Versions and configuration

Active World Mods

> Dark Days Ahead
> Alternative Map Key
> No Rail Stations
> Graphical Overmap
> sees-player-icon, HitButton_iso
> sees-player icon, retrodays
> Magiclysm
> Graphical Overmap Magiclysm

# Additional context

> Ungenomoldic  Today at 6:23 PM
> Yeah, that error is everywhere in my game
> ''Item stanag_mag_10 was loaded with charges, but can not have any!''
> ''Item mask_gas was loaded with charges, but can not have any!''
> ''Item inhaler was loaded with charges, but can not have any!''
> You name it, there is a ''was loaded with charges!'' for it

### LINKED ISSUE #41907: DEBUG: Item mask_bunker was loaded with charges, but can not have any!
# Describe the bug
 FUNCTION : void item::io(Archive&) [with Archive = io::JsonObjectInputArchive]
 FILE     : src/savegame_json.cpp
 LINE     : 2390

# Steps To Reproduce
Unknown occurred while autosaving.

# Expected behavior
?

# Screenshots
N/R

# Versions and configuration
- OS: Linux
    - OS Version: LSB Version: core-9.20170808ubuntu1-noarch:printing-9.20170808ubuntu1-noarch:security-9.20170808ubuntu1-noarch; Distributor ID: Ubuntu; Description: Ubuntu 18.04.4 LTS; Release: 18.04; Codename: bionic; 
- Game Version: 0.E-3780-g54cde4b [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Monsters [No_Fungi],
    Stats Through Skills [StatsThroughSkills]
]

# Additional context
debug.log
[debug.log](https://github.com/CleverRaven/Cataclysm-DDA/files/4881526/debug.log)

Save file
[Bull Shoals.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/4881518/Bull.Shoals.zip)

### LINKED ISSUE #42081: Battery UPS bug
DEBUG    : Item battery_ups was loaded with charges, but can not have any!

 FUNCTION : void item::io(Archive&) [with Archive = io::JsonObjectInputArchive]
 FILE     : src/savegame_json.cpp
 LINE     : 2363

windows 10 
Version game stable E.2

image

### PR DESCRIPTION
**Pull request by wapcaplet**
#### Summary

SUMMARY: Bugfixes "Fix item loaded with charges but cannot have any error"

#### Purpose of change

Prevent this error from appearing for a lot of items: `Item ... was loaded with charges, but can not have any`

Fix #41821 
Fix #41907 (belatedly)

#### Describe the solution

Add obsolete no-longer-charged items to `data/json/obsoletion/charge_removal.json`, including:

- acoustic_guitar
- anesthetic kit
- bagpipes
- banjo
- bone_flute
- clarinet
- flute
- guitar_electric
- handflare
- handflare_lit
- handflare_dead
- pseudo_bio_picklock (fixed spelling)
- rx11_stimpack
- saxophone
- sewing_kit
- trumpet
- ukulele
- violin
- violin_golden
- acoustic_guitar
- bagpipes
- guitar_electric
- harmonica_holder
- saxophone
- tuba
- whistle
- battery_ups (see "Additional context")

Removed `initial_charges` and `max_charges` from all masks and scuba tanks, and added them to the charge-obsoletion list as well:

- mask_bunker
- mask_filter
- mask_fsurvivor
- mask_fsurvivorxl
- mask_gas
- mask_gas_xl
- mask_h20survivor
- mask_h20survivorxl
- mask_hsurvivor
- mask_lsurvivor
- mask_ski
- mask_ski_loose
- mask_survivor
- mask_survivorxl
- mask_wsurvivor
- mask_wsurvivorxl
- rebreather
- rebreather_xl
- scuba_tank
- small_scuba_tank
- handflare
- rx11_stimpack

Savegames containing any of the above items should no longer have "charges" incorrectly saved in future savegames.

Also alphabetize the items in the blacklist file for ease of future modification, as still more items are likely to have their charges removed.

#### Describe alternatives you've considered

Considered adding other items that were known to be affected by this (such as acoustic_guitar, harmonica_holder, mask_filter and so on) but I can no longer reproduce the bug with those items and would have no way to verify it.

#### Testing

1. Debug-spawn a pile of 1 of every item (press `E` from the debug-spawn-item menu)
2. Teleport 9 overmap squares away
3. Save game, quit and reload
4. Walk toward the pile of items - "loaded with charges but can not have any" errors will occur unles items are in the `charge_removal.json` list.

Or use this save game with a pile already set up about 9 overmap tiles to the left (near the evac shelter). 

[no-charges-big-pile.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5081593/no-charges-big-pile.zip)

Walk westward until errors appear (or don't).

For the de-charged gas masks:
1. Spawn some gas masks from the debug menu, along with a smoke bomb
2. Activate and wear one of the masks - it should be active and yellow
3. Activate and drop the bomb
4. Stand there and breathe the smoke a while
5. Make sure the filter is depleted

image

#### Additional context

This error message was originally added in #32466 to catch the possibility of items being erroneously saved with "charges" when they should not have any, as they would prevent items from stacking correctly.

Item `battery_ups` was reported with this error in #42081 (and in closed duplicates #39434 and #39591). I can't find any evidence that this item ever had charges, but apparently a few people had their savegames messed up with "charges" for this item. Incorrectly having charges may still be an issue for this item in some savegames, so I have added `battery_ups` to the blacklist as well, simply to prevent showing the error (although this PR does not claim to have fixed any of those three issues).

Many thanks to @anothersimulacrum and @KorGgenT for their patience in working through this with me on Discord.

### CODE REVIEW

**Comment by jbytheway** [comment]
Do flares (specifically `handflare`) work after this change?  Has flare timing been changed to work via some means other than charges?

**Comment by wapcaplet** [comment]
> Do flares (specifically `handflare`) work after this change? Has flare timing been changed to work via some means other than charges?

@jbytheway I have no idea. Should it be? My interpretation was that the item simply needed to be blacklisted so it won't get this error.

**Comment by jbytheway** [comment]
Some items needed further changes, e.g. how lighters now contain butane instead of having charges; these might have similar needs. I'm not familiar enough with the details to be sure.

**Comment by wapcaplet** [comment]
I have seen (and seen reports of) this error occurring with `harmonica_holder`, `acoustic_guitar`, `jumper_cable`, `mask_filter`, `mask_bunker`, `banjo` and other items. However I have not determined steps to reproduce these errors, so have not included them yet.

**Comment by mqrause** [comment]
> Do flares (specifically `handflare`) work after this change? Has flare timing been changed to work via some means other than charges?

Seems like they don't have charges but ammo now.

I'm not sure how all of this works, but I would guess it's an issue with saving the items rather than loading them. This isn't an old save that needs to be fixed after all.

File: data/json/obsoletion/charge_removal.json [hunk 1]
```diff
@@ -65,7 +73,7 @@
       "portable_game",
       "powered_earmuffs",
       "powered_earmuffs_on",
-      "pseuso_bio_picklock",
+      "pseudo_bio_picklock",
```
Comment (jbytheway) [inline data/json/obsoletion/charge_removal.json]: This typo suggests that we ought to add some code somewhere that verifies the ids in this list are valid.

Comment (wapcaplet) [reply data/json/obsoletion/charge_removal.json]: I guess somewhe in `src/init.cpp:DynamicDataLoader::check_consistency` would be suitable. But if this list is just a temporary way to ignore charge errors with these items when loaded from an older save game, I don't foresee much need to add more to it after this PR, and suspect the whole list could be eliminated some time after 0.F.

Comment (wapcaplet) [reply data/json/obsoletion/charge_removal.json]: A couple more hours of immersion in this bug have brought me many doubts about my statement above, but either way it's not something I'm prepared to do for this PR.

**Comment by wapcaplet** [comment]
Related mainline commits that have "fixed" erroneously charged items:
- 94a4df7eb7 - Remove max_charges from musical instruments violin, tuba, flute, banjo, guitars and others
- da256333b - Remove charges entirely from musical instruments

Removed max_charges but left initial_charges (should have removed both):
- 61a16bb3b - Allowing gas mask, rebreather, scuba tank etc. to be reloaded

**Comment by Marrim** [comment]
I'm glad this is resolved, but handflares still spawn with 0 charge (of magnesium) and are thus unusable.

**Pull request merged**